### PR TITLE
Handle default package in the zero-config Compact serializers better for Java 8

### PR DESCRIPTION
**Pull request by mdumandag**
In Java 9 and above, for the classes in the default package, the
`clazz.getPackage()` method returns a non-null package with an
empty class name.

However, it returns a null package in Java 8, for the same classes.

We were saying that a class is not-zero-config-serializable if the
package was null (because it is null if it is a primitive type, array
or void in Java 9+, which are not supported in the paths that we do
the checks in this method). That was not correct in Java 8.

Now, if the package is null, we are performing the checks done in
Java 9+ and above again to distinguish classes in the default package.

### CODE REVIEW

**Review by ihsandemir — COMMENTED** [review]
can you add a test to test for Java 8 and another one for 9+versions?

**Comment by mdumandag** [comment]
@ihsandemir I can't do that because you cannot import classes from the default package in other packages

**Comment by ihsandemir** [comment]
> @ihsandemir I can't do that because you cannot import classes from the default package in other packages

don't we have Java 8 and Java 9 nightly builds? We can easily have a test and we make sure that it works in both builds correctly.

**Comment by mdumandag** [comment]
@ihsandemir like I write above, the problem is not about Java versions, etc. The thing is, if I have the class `Foo` in the default package (under the java directory), I can't import it in the tests that have another package like `com.hazelcast.internal.serialization.impl.compact.CompactSerializationTest`. There is no way to access `Foo` in any package other than the default package.

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactUtil.java [hunk 1]
```diff
@@ -215,7 +215,24 @@ public static void verifyFieldClassShouldBeSerializedAsCompact(CompactStreamSeri
...
+            // We won't hit this line for primitive types, as they are already
+            // supported by the reflective serializers. For arrays, we can only
+            // hit this line for arrays of arrays which is not supported by
+            // Compact. And, Void is not a type we support in the Compact
+            // serialization.
+
+            // So, we can return false here, if the clazz is one of the types
+            // mentioned above, and return true if not, to be able to serialize
+            // user classes without a package in Java 8.
+            return !clazz.isPrimitive() && !clazz.isArray() && !Void.class.equals(clazz);
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactUtil.java:235]: what about types like Time, Date, etc. which are compact serializable?

Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactUtil.java:235]: What about them? We won't be hitting this line for any of those classes. This is for nested compact objects and arrays of arrays basically

Comment (mdumandag) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactUtil.java:235]: @ihsandemir I have added tests for array of array and void field types, and this is the best we can test

**Comment by mdumandag** [comment]
Unrelated test failure reported on https://github.com/hazelcast/hazelcast/issues/18644

**Comment by mdumandag** [comment]
one more unrelated failure reported on https://github.com/hazelcast/hazelcast/issues/22205

**Comment by mdumandag** [comment]
run-lab-run

**Review by ihsandemir — APPROVED** [review]

**Comment by mdumandag** [comment]
I am quite unlucky today :/ Another unrelated failure I reported on https://github.com/hazelcast/hazelcast/issues/22206. I will run the tests again, and if it fails again, we can force merge the PR

**Comment by mdumandag** [comment]
run-lab-run

**Pull request merged**